### PR TITLE
Added Decompose example for native iOS navigation

### PR DESCRIPTION
**Pull request by ttypic**
Fooling around native iOS navigation, trying to make it friendly with Decompose using UIKit interop 

It's relative to [discussion](https://github.com/arkivanov/Decompose/discussions/76)

### CODE REVIEW

**Comment by arkivanov** [comment]
Hey! Thanks for the example, this looks pretty useful! However, I realise that the overall sample apps gets bloated and difficult to read. What do you think about the following improvement? 

Currently the Counters tab demonstrates having multiple stacks in the same component - `CountersComponent` contains two stacks. We could sacrifice this in favour of a simple stack navigation with toolbars/navbars.

Does is sound reasonable? I could refactor the existing samples into the new structure, and you could add your improvement for the iOS app on top of that.

**Comment by arkivanov** [comment]
Also maybe we could enable the native swipe back gesture in this case?

**Comment by ttypic** [comment]
I am 100% agree with you, that's a good plan

**Comment by arkivanov** [comment]
> I am 100% agree with you, that's a good plan

I have updated the Counter sample - #269. Feel free to apply your changes on top of the new code.

**Comment by ttypic** [comment]
I've updated the code, but it is not yet ready. Basically all the functionality works: swipe gesture, back button, etc. But there are two issues:

- I can't find a way to create clean data-driven component with native iOS Navigation. It's really hard to somehow intercept back button logic. So I just detect that view did disappear and invoke callback to sync changes. As I understand new iOS 16 `NavigationStack` has the same problem with `path` property, even when I send constant Binding, back button still works
- Another issue is with title, navigation title appears with small delay on push

I am trying to figure out how to fix this, especially title, if you have any ideas they be very welcome. 

Also I haven't yet cleaned up the code.

**Comment by arkivanov** [comment]
Thanks for the updates! I'm not quite sure, but the issue with title appears to happen when SwiftUI is mixed with UIKit - https://stackoverflow.com/questions/74247582/navigation-bar-title-appears-with-delay-in-mixed-uikit-and-swiftui-project. Maybe we could try the new `NavigationStack`? Or maybe try the workaround mentioned in SOF?

**Comment by ttypic** [comment]
@arkivanov Looks like it works! Both on new iOS 16 with brand-new `NavigationStack` and legacy realization with UIKit Introp for  XCode < 14 and iOS < 16 (code looks much uglier when NavigationStack but UI feels exactly the same)

File: sample/app-ios/app-ios/CountersView.swift [hunk 1]
```diff
@@ -22,7 +25,19 @@ struct CountersView: View {
     }
 
     var body: some View {
-        CounterView(childStack.value.active.instance)
+        if #available(iOS 16, *) {
+            let pathBinding = Binding(get: { componentIndexes.dropFirst() }, set: { _ in activeChild.onPrevClicked() })
+            NavigationStack(path: pathBinding) {
+                CounterView(components[0])
+                        .navigationDestination(for: Int.self) {
```
Comment (arkivanov) [inline sample/app-ios/app-ios/CountersView.swift]: Not quite sure, but maybe we should use some kind of stable keys here? E.g. for cases when the active component is replaced with another one. Maybe configuration `hashCode()` as a string? Or just `hashCode()` as int?

File: sample/app-ios/app-ios/CountersStackView.swift [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                    viewControllers[i] = parent.createViewController(parent.components[i])
+                }
+            }
+
+            preservedComponents = parent.components
+        }
+    }
+
+    class NavigationItemHostingController<Content: View>: UIHostingController<Content> {
+        fileprivate(set) var stackView: CountersStackView?
```
Comment (arkivanov) [inline sample/app-ios/app-ios/CountersStackView.swift]: Shouldn't this one also be `weak`?

File: sample/app-ios/app-ios/CountersView.swift (hunk 1 above)
Comment (ttypic) [reply sample/app-ios/app-ios/CountersView.swift]: Totally agree, fixed this

**Review by arkivanov — APPROVED** [review]

**Comment by arkivanov** [comment]
Thank you!

**Pull request merged**